src/main/java/io/cdap/plugin/sfmc/connector/MarketingConnectorConfig.java

```diff
@@ -0,0 +1,150 @@
+    this.clientSecret = clientSecret;
+    this.authEndpoint = authEndpoint;
+    this.soapEndpoint = soapEndpoint;
+  }
+
+  public String getClientId() {
+    return clientId;
+  }
+
+  public String getClientSecret() {
+    return clientSecret;
+  }
+
+  public String getAuthEndpoint() {
+    return authEndpoint;
+  }
+
+  public String getSoapEndpoint() {
+    return soapEndpoint;
+  }
+
+  /**
+   * validates all the fields which are mandatory for the connection.
+   */
+  public void validateCredentials(FailureCollector collector) {
+    if (!shouldConnect()) {
+      return;
+    }
+    if (Util.isNullOrEmpty(clientId)) {
+      collector.addFailure("Client ID must be specified.", null)
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_CLIENT_ID);
+    }
+    if (Util.isNullOrEmpty(clientSecret)) {
+      collector.addFailure("Client Secret must be specified.", null)
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_CLIENT_SECRET);
+    }
+    if (Util.isNullOrEmpty(authEndpoint)) {
+      collector.addFailure("Auth Endpoint  must be specified.", null)
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_AUTH_API_ENDPOINT);
+    }
+    if (Util.isNullOrEmpty(soapEndpoint)) {
+      collector.addFailure("Soap Endpoint must be specified.", null)
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_SOAP_API_ENDPOINT);
+    }
+    collector.getOrThrowException();
+    validateSalesforceConnection(collector);
+  }
+
+  public void validateSalesforceConnection(FailureCollector collector) {
+
+    try {
+      MarketingCloudClient.create(getClientId(), getClientSecret(),
+        getAuthEndpoint(),
+        getSoapEndpoint());
+    } catch (ETSdkException e) {
+      collector.addFailure("Unable to connect to Salesforce Instance.",
+          "Ensure properties like Client ID, Client Secret, API Endpoint " +
+            ", Soap Endpoint, Auth Endpoint are correct.")
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_CLIENT_ID)
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_CLIENT_SECRET)
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_AUTH_API_ENDPOINT)
+        .withConfigProperty(MarketingCloudConstants.PROPERTY_SOAP_API_ENDPOINT)
+        .withStacktrace(e.getStackTrace());
+    }
+  }
+
+  /**
+   * Returns true if any of the required parameters are macro enabled
+   */
+  public boolean shouldConnect() {
+    return !containsMacro(MarketingCloudConstants.PROPERTY_CLIENT_ID) &&
+      !containsMacro(MarketingCloudConstants.PROPERTY_CLIENT_SECRET) &&
+      !containsMacro(MarketingCloudConstants.PROPERTY_SOAP_API_ENDPOINT) &&
+      !containsMacro(MarketingCloudConstants.PROPERTY_AUTH_API_ENDPOINT);
+  }
+
+  public Schema getSchema(SourceObject sourceObject) {
+    try {
+      MarketingCloudClient client = MarketingCloudClient.create(this.getClientId(),
+        this.getClientSecret(),
+        this.getAuthEndpoint(),
+        this.getSoapEndpoint());
```

Comment:
Created a function 'getInstance' in MarketingCloudUtil class and accessing the function wherever it is required to avoid multiple calls to create ETClient object.

Reply:
MarketingCloudUtil is not the correct place for this method. Instead keep the method in the MarketingCloudClient and rename it to getOrCreate.

Reply:
moved it to MarketingCloudClient and changed the function name to getOrCreate.